data-prepper-plugins/obfuscate-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/obfuscation/ObfuscationProcessor.java

```diff
@@ -0,0 +1,191 @@
+        Matcher matcher = pattern.matcher(rawPattern);
+        if (matcher.matches()) {
+            try {
+                CommonPattern cp = CommonPattern.valueOf(matcher.group(1));
+                // If found, used the common pattern regex
+                this.pattern = cp.getExpr();
+            } catch (IllegalArgumentException e) {
+                LOG.error("Unable to find a predefined pattern for {}", matcher.group(1));
+                // Throw exception as the pattern is for sure invalid
+                throw new InvalidPluginConfigurationException("Unable to find a predefined pattern for \"" + rawPattern + "\".");
+            }
+        } else {
+            // Try to validate the custom pattern
+            try {
+                Pattern.compile(rawPattern);
+                this.pattern = rawPattern;
+            } catch (PatternSyntaxException e) {
+                LOG.error(e.getMessage());
+                throw new InvalidPluginConfigurationException("Invalid Pattern: \"" + rawPattern + "\" for Key " + this.key);
+            }
+        }
+        LOG.debug("Creating Processor to obfuscate field {} with pattern {}", this.key, this.pattern);
+
+    }
+
+    @Override
+    public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
+
+        for (final Record<Event> record : records) {
+            final Event recordEvent = record.getData();
+
+            if (!recordEvent.containsKey(key)) {
+                continue;
+            }
+
+            String rawValue = recordEvent.get(key, String.class);
+            String newValue = obfuscate(rawValue);
+
+            // Update the event record.
+            if ("".equals(newKey)) {
+                recordEvent.put(key, newValue);
+            } else {
+                recordEvent.put(newKey, newValue);
+            }
+        }
+        return records;
+    }
+
+
+    /**
+     * Returns the last {unObfuscatedLength} of characters of a string {str}
+     * <p>For example:</p>
+     * <ul>
+     * <li>getUnObfuscatedString("Hello", 3) = "llo"</li>
+     * <li>getUnObfuscatedString("Hello", 10) = "Hello"</li>
+     * </ul>
+     */
+    private String getUnObfuscatedString(String str, int unObfuscatedLength) {
+        int startIndex = Math.max(str.length() - unObfuscatedLength, 0);
+        return str.substring(startIndex);
+    }
+
+
+    private String obfuscate(String str) {
+
+        if ("".equals(pattern)) {
+            // This is to replace the whole field.
+            return baseReplacement + getUnObfuscatedString(str, unObfuscatedLength);
+        }
+
+        if (unObfuscatedLength == 0) {
+            // Don't need to calculate the replaced string
+            // Simply use the replaceAll function with baseReplacement string.
+            return str.replaceAll(pattern, baseReplacement);
+        }
+
+        // Try to find all matches and then replace one by one.
+        Pattern p = Pattern.compile(pattern);
```

Comment:
Can we only compile the patterns once at startup in the constructor, and then save the compiled patterns for reuse here to improve performance. Compiling the pattern every time will have worse performance